src/NServiceBus.Core/Routing/EndpointInstances.cs

```diff
@@ -2,106 +2,67 @@ namespace NServiceBus.Routing
 {
     using System;
     using System.Collections.Generic;
-    using System.Threading.Tasks;
+    using System.Linq;
 
     /// <summary>
     /// Stores the information about instances of known endpoints.
     /// </summary>
     public class EndpointInstances
     {
-        internal Task<IEnumerable<EndpointInstance>> FindInstances(string endpoint)
+        internal IEnumerable<EndpointInstance> FindInstances(string endpoint)
         {
-            HashSet<EndpointInstance> staticInstances;
-            staticRules.TryGetValue(endpoint, out staticInstances);
-
-            if (dynamicRules.Count > 0)
-            {
-                return FindDynamicInstances(endpoint, staticInstances);
-            }
-
-            if (staticInstances != null)
-            {
-                return Task.FromResult<IEnumerable<EndpointInstance>>(staticInstances);
-            }
-
-            return Task.FromResult<IEnumerable<EndpointInstance>>(new[]
-            {
-                new EndpointInstance(endpoint)
-            });
+            HashSet<EndpointInstance> instances;
+            return cache.TryGetValue(endpoint, out instances)
+                ? instances
+                : EnumerableEx.Single(new EndpointInstance(endpoint));
         }
 
-        async Task<IEnumerable<EndpointInstance>> FindDynamicInstances(string endpoint, HashSet<EndpointInstance> staticInstances)
+        internal void SetLogChangeAction(Action<string> logChangeAction)
         {
-            var dynamicInstances = staticInstances != null ? new HashSet<EndpointInstance>(staticInstances) : new HashSet<EndpointInstance>();
-            foreach (var rule in dynamicRules)
-            {
-                var instancesFromRule = await rule.Invoke(endpoint).ConfigureAwait(false);
-                foreach (var instance in instancesFromRule)
-                {
-                    dynamicInstances.Add(instance);
-                }
-            }
-
-            if (dynamicInstances.Count == 0)
-            {
-                dynamicInstances.Add(new EndpointInstance(endpoint));
-            }
-
-            return dynamicInstances;
+            this.logChangeAction = logChangeAction;
         }
 
         /// <summary>
-        /// Adds a dynamic rule for determining endpoint instances.
+        /// Adds or replaces a set of endpoint instances registered under a given key (registration source ID).
         /// </summary>
-        /// <param name="dynamicRule">The rule.</param>
-        public void AddDynamic(Func<string, Task<IEnumerable<EndpointInstance>>> dynamicRule)
+        /// <param name="sourceKey">Source key.</param>
+        /// <param name="endpointInstances">List of endpoint instances known by this source.</param>
+        public void AddOrReplaceInstances(object sourceKey, IList<EndpointInstance> endpointInstances)
```

Comment:
since it's a public API I think it may not be very obvious what we expect and since object equality can differ a lot, I think a string parameter would make it easier to use.